include/sys/onoff.h

```diff
@@ -0,0 +1,533 @@
+ * onoff_release() through the provided signal.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param sigp pointer to the signal to use for notification.  The
+ * value must not be null.  The signal must be reset before the client
+ * object is passed to the on-off service API.
+ *
+ * @retval 0 on success
+ * @retval -EINVAL if cli is null
+ * @retval -EINVAL if sigp is null
+ */
+static inline int onoff_client_init_signal(struct onoff_client *cli,
+					   struct k_poll_signal *sigp)
+{
+	if ((cli == NULL) || (sigp == NULL)) {
+		return -EINVAL;
+	}
+
+	*cli = (struct onoff_client){
+		.async = {
+			.signal = sigp,
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_SIGNAL,
+	};
+
+	return 0;
+}
+
+/**
+ * @brief Initialize an on-off client to notify with a callback.
+ *
+ * Clients that use this initialization will by notified of the
+ * completion of operations submitted through on-off service API
+ * through the provided callback.  Note that callbacks may be invoked
+ * from various contexts; see onoff_client_callback.
+ *
+ * @param cli pointer to the client state object.
+ *
+ * @param cb a function pointer to use for notification.
+ *
+ * @retval 0 on success
+ * @retval -EINVAL if cli is null
+ * @retval -EINVAL if cb is null
+ */
+static inline int onoff_client_init_callback(struct onoff_client *cli,
+					     onoff_client_callback cb)
+{
+	if ((cli == NULL) || (cb == NULL)) {
+		return -EINVAL;
+	}
+
+	*cli = (struct onoff_client){
+		.async = {
+			.callback = cb,
+		},
+		.flags = ONOFF_CLIENT_NOTIFY_CALLBACK,
+	};
+
+	return 0;
+}
+
+/**
+ * @brief Request a reservation to use an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to request the resource be made available.
+ * If initiation of the operation succeeds the result of the request
+ * operation is provided through the configured client notification
+ * method, possibly before this call returns.
+ *
+ * Note that the call to this function may succeed in a case where the
+ * actual request fails.  Always check the operation completion
+ * result.
+ *
+ * This function is isr-capable, and may suspend unless client
+ * notification specifies no-wait.
+ *
+ * @param srv the service that will be used.
+ *
+ * @param cli a non-null pointer to client state providing
+ * instructions on synchronous expectations and how to notify the
+ * client when the request completes.  Behavior is undefined if client
+ * passes a pointer object associated with an incomplete service
+ * operation.
+ *
+ * @retval Non-negative on successful (initiation of) request
+ * @retval -EIO if service has recorded an an error
+ * @retval -EINVAL if the parameters are invalid
+ * @retval -EAGAIN if the reference count would overflow
+ * @retval -EWOULDBLOCK if a non-blocking request was made and could
+ *         not be satisfied
+ */
+__syscall int onoff_request(struct onoff_service *srv,
```

Comment:
When "signal type" client needs to call `onoff_service_has_error()` after `onoff_request()`? Only when `-EIO` is returned?

Reply:
When `onoff_request()` returns `-EO` it means the service already had an error at the point the call was made; there's no need to check has_error because it will return true (unless something slips in and resets).

If the request produces an error, that's indicated in the operation result (for no-wait, signal, or callback).  There the negative result indicates a failure to complete the transition, and any error puts the service into an error state (because it can't tell if the failure left the system in the original state or an undefined state).